consumer/consumererror/error.go

```diff
@@ -0,0 +1,178 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package consumererror // import "go.opentelemetry.io/collector/consumer/consumererror"
+
+import (
+	"errors"
+
+	"google.golang.org/grpc/status"
+
+	"go.opentelemetry.io/collector/consumer/consumererror/internal/statusconversion"
+)
+
+// Error acts as a container for errors coming from pipeline components.
+// It may hold multiple errors from downstream components, and is designed
+// to act as a way to accumulate errors as it travels upstream in a pipeline.
+// `Error` should be obtained using `errors.As` and as a result, ideally
+// a single instance should exist in an error stack. If this is not possible,
+// the most `Error` object should be highest on the stack.
+//
+// Experimental: This API is at the early stage of development and may change without backward compatibility
+type Error struct {
+	errors []ErrorData
+}
+
+// ErrorData is intended to be used to encapsulate various information
+// that can add context to an error that occurred within a pipeline component.
+// ErrorData objects are constructed through calling `New` with the relevant
+// options to capture data around the error that occurred. It can then be pulled
+// out by an upstream component by calling `Error.Data`.
+//
+// Experimental: This API is at the early stage of development and may change without backward compatibility
+type ErrorData interface {
+	Error() string
+
+	Unwrap() error
+
+	// Returns a status code and a boolean indicating whether a status was set
+	// by the error creator.
+	HTTPStatus() (int, bool)
+
+	// Returns a status code and a boolean indicating whether a status was set
+	// by the error creator.
+	GRPCStatus() (*status.Status, bool)
+
+	// Disallow implementations outside this package to allow later extending
+	// the interface.
+	unexported()
+}
+
+type errorData struct {
+	error
+	httpStatus *int
+	grpcStatus *status.Status
+}
+
+// ErrorOption allows annotating an Error with metadata.
+type ErrorOption func(error *errorData)
```

Comment:
Not recommended to have a public type that wraps a private type. I think the best way to do this is to use the interface pattern, see https://github.com/open-telemetry/opentelemetry-collector/blob/ecbe02ed685a6f662915f781c2646adeafd27194/receiver/internal/factory.go#L60

Reply:
I think I like this a bit better, thanks. Implemented.